pkg/models/host.go

```diff
@@ -86,43 +78,94 @@ func (h *Host) Connected() bool {
 }
 
 func (h *Host) Disconnect() error {
-	if !h.Connected() {
+	if h.Connection == nil {
 		return nil
 	}
-	return h.Connection.Client.Close()
+	err := h.Connection.Close()
+	return err
 }
 
 // Upload uploads a local file to a remote file on the host.
-func (h *Host) Upload(localFile string, remoteFile string) error {
+func (h *Host) Upload(localFile string, remoteFile string, timeout time.Duration) error {
 	if !h.Connected() {
-		if err := h.Connect(constants.SSHFileOpsTimeout); err != nil {
+		if err := h.Connect(); err != nil {
 			return err
 		}
 	}
-	return h.Connection.Client.Upload(localFile, remoteFile)
+	var err error
+	ch := make(chan struct{})
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+	go func() {
+		err = h.Connection.Upload(localFile, remoteFile)
+		close(ch)
+	}()
+	select {
+	case <-ctx.Done():
+		return fmt.Errorf("upload timeout of %d seconds for host %s", uint(timeout.Seconds()), h.IP)
+	case <-ch:
+	}
+	return err
 }
 
 // Download downloads a file from the remote server to the local machine.
-func (h *Host) Download(remoteFile string, localFile string) error {
+func (h *Host) Download(remoteFile string, localFile string, timeout time.Duration) error {
 	if !h.Connected() {
-		if err := h.Connect(constants.SSHScriptTimeout); err != nil {
+		if err := h.Connect(); err != nil {
 			return err
 		}
 	}
 	if err := os.MkdirAll(filepath.Dir(localFile), os.ModePerm); err != nil {
 		return err
 	}
-	return h.Connection.Client.Download(remoteFile, localFile)
+	var err error
```

Comment:
convert this into func like runWithTimeout where you pass a func and wrap it around context and channel to keep code DRY